Lib/csharp/std_optional.i

```diff
@@ -0,0 +1,329 @@
+//    This macro is used to bind arithmetic types (int, float, double, etc.) to C# nullable value type (represented as Nullable<T>).
+//    Example:
+//    %optional_arithmetic(double, OptDouble)
+//    This will generate code to handle std::optional<double>.
+//    An internal class name OptDouble will be created to bind the optional type from C++ to C# nullable value type. The internal class will be marked as internal by default.
+//    If you want to change the access modifier of the internal class for one specific type, you can pass an optional CLASSMODIFIER parameter to the macro.
+//    See the SWIG_STD_OPTIONAL_INTERNAL_CLASS_MODIFIER macro for more information.
+// 
+// 2. %optional_string():
+//    This macro is used specifically for std::optional<std::string>.
+//    Example:
+//    %optional_string()
+//    This will generate code to handle std::optional<std::string>.
+//
+// 3. %optional(TYPE):
+//    This macro is used to bind structs and classes to either C# nullable reference type (if SWIG_STD_OPTIONAL_USE_NULLABLE_REFERENCE_TYPES is defined) or C# non-nullable reference type.
+//    Example:
+//    %optional(MyStruct)
+//    This will generate code to handle std::optional<MyStruct>.
+//
+// The SWIG_STD_OPTIONAL_DEFAULT_TYPES macro is used to define a set of default types that are supported by SWIG when working with std::optional.
+// This macro ensures that common types such as int, double, and std::string are automatically recognized and handled by SWIG when generating bindings.
+//
+// The SWIG_STD_OPTIONAL_USE_NULLABLE_REFERENCE_TYPES macro definition is used to enable the use of nullable reference types in the generated SWIG bindings.
+// When this macro is defined, SWIG will generate code that treats std::optional types as C# nullable reference types, allowing the compiler to run null-state analysis (C# >= 8 required).
+// When this macro is defined, any module using this file should be declared with the #nullable enable directive.
+// Example:
+// %module(csbegin="#nullable enable\n") MyModule
+//
+// The SWIG_STD_OPTIONAL_INTERNAL_CLASS_MODIFIER macro is used to define the access modifier for the internal class generated for the arithmetic optional types. By default it is set to internal.
+// You may want to define it to 'public' so the internal class is accessible from outside the assembly it's defined in. See the documentation of the SWIG_CSBODY_PROXY macro for more information.
+// Defining this macro will affect all optional arithmetic types defined in the module.
+
+#if defined(SWIG_STD_OPTIONAL_USE_NULLABLE_REFERENCE_TYPES)
+%define SWIG_STD_OPTIONAL_NULLABLE_TYPE "?" %enddef
+#else
+%define SWIG_STD_OPTIONAL_NULLABLE_TYPE "" %enddef
+#endif
+
+#ifndef SWIG_STD_OPTIONAL_STRINGIFY
+#define SWIG_STD_OPTIONAL_STRINGIFY_(x) #x
+#define SWIG_STD_OPTIONAL_STRINGIFY(x) SWIG_STD_OPTIONAL_STRINGIFY_(x)
+#endif
+
+
+%{
+    #include <optional>
+%}
+
+%include <stdint.i>

```

Comment:
Implicitly including `stdint.i` opens something of a can of worms.  By default SWIG aims to generate code independent of the platform it is run on, which can then be built anywhere.  A few places violate this, and `stdint.i` is one of them as the code generated is tailored to the system it will be built on (the problem is that for most target languages SWIG needs to work in terms of `int`, `long`, `long long`, etc which of those e.g. `int64_t` corresponds to depends on the platform).

So I think it would be better to avoid doing this by default.

Reply:
Oh dear I didn't get the notification on April 2nd when Olly reviewed the PR. I can try without the inclusion and see if that works. I've been using this code for more than a year in multiple projects now.